Implemented Drag and Drop for Ozone/Wayland

jkim-julie opened the pull request:
It introduces DesktopDragDropClientOzone and some APIs related to DnD
to PlatformWindowDelegate to deliver DnD status from platform and
system. OnDragEnter/OnDragMotion/OnDragDrop/OnDragLeave reports DnD
status with information based on the status. OnDragDataReceived is
called when it gets dragged data from system, OnDragDataRequest is
called when system requests dragged data and OnCloseDrag is called
when DnD is canceled or performed. These APIs are connected to
DesktopDragDropClientOzone.

It exteneds WaylandDataDevice, WaylandDataSource and WaylandDataOffer
for DnD on Wayland.

The working sequence is blow.
1) When it gets dragging.
WaylandDataDevice gets OnEnter() and DesktopDragDropClientOzone set
dragged data if it has |source_data_|. It means dragging is started
internally. After that, it continously gets Motion and Drop events.
When it gets OnDragDrop, it tries to read dragged data if it doesn't
have yet. As OnDragLeave event comes right after OnDragDrop, it
defers OnDragLeave event handling. After finishing it, it calls
callback and OnDragLeave is handled. The reason that it reads
dragged data on OnDragDrop event, it's blocked if |fd| offered by
|drag_offer_| doesn't have data yet.

2) When it sends dragging.
It calls 'wl_data_device_start_drag' through
'DesktopDragDropClientOzone::StartDragAndDrop'. After getting
OnDragEnter/Motion, it gets WaylandDataSource::OnSend() and fills
|fd| with the dragged data. If Drop is performed without problem,
it gets OnDnDDropPerformed() and OnDnDFinished(). Otherwise,
it gets OnCancel().

jkim-julie commented:
Hi, @msisov  and @tonikitoo, could you take a look at it?

It has several TODOs still and some points I'd like you to look into carefully.

1) I set 'wl_data_device_manager' version |wl_registry_listener| gets, it is '3' on my environment. It was previously '1'. When I tried to version '1', it didn't call OnDrgDrop.

2) TODO: CreateSHMBuffer() should be shared. I'll continue to work on it.

3) As I wrote on description, Current Chromium implementation requires dragged data on OnDragEnter and it's not updated until OnDragDrop. I tried to read drag data on OnDragEnter but it blocks a whole sequence because |fd| offered by |drag_offer_| doesn't have data yet. For this issue, I looked into all ways to avoid it as far as I can imagine. After that, I looked into Weston code and I saw that Weston also reads dragged data on OnDragDrop, https://github.com/wayland-project/weston/blob/280b730144cafc58dc78ac60b3a3dbe61b6b6b50/clients/dnd.c#L754.
BTW, I found that it has shortcut solution on AuraX11, https://cs.chromium.org/chromium/src/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.cc?q=desktop_drag_&sq=package:chromium&g=0&l=344.
So, I mimicked shortcut if it has |source_data_| and read dragged data only when the dragging started from other applications.

4) I tried to access to DragDropClient from Wyaland platform layer directly but I couldn't as platform layer knows only WaylandWindow. If it knows aura::Window, it could get DragDropClient through GetDragDropClient(Window* root_window) and it could make code simpler.

5) WaylandPointer::Button isn't called with WL_POINTER_BUTTON_STATE_RELEASED after drop action happens. https://cs.chromium.org/chromium/src/ui/ozone/platform/wayland/wayland_pointer.cc?q=wayland_pointer.cc&sq=package:chromium&g=0&l=105.
So, It keeps captured state. I didn't looked into it deeply. I'll keep updating.

6) TODO: Make unittests. I'll continue to work on it.

msisov commented on ui/ozone/platform/wayland/wayland_connection.cc:
```diff
@@ -325,6 +325,31 @@ bool WaylandConnection::IsSelectionOwner() {
+                                  const SkBitmap& bitmap) {
+  if (!data_source_) {
+    wl_data_source* data_source = data_device_manager_->CreateSource();
+    drag_data_source_.reset(new WaylandDataSource(data_source));
```
std::make_unique\<type\>

msisov replied:
Btw, can you move creation of the WaylandDataSource into a separate method and share it with the OfferClipboardData? Maybe, set_connection can also be removed and WaylandConnection can be passed as an argument in the construction parameters.

msisov commented on ui/ozone/platform/wayland/wayland_connection.cc:
```diff
@@ -489,7 +514,7 @@ void WaylandConnection::Global(void* data,
              strcmp(interface, "wl_data_device_manager") == 0) {
     wl::Object<wl_data_device_manager> data_device_manager =
-        wl::Bind<wl_data_device_manager>(registry, name, 1);
+        wl::Bind<wl_data_device_manager>(registry, name, version);
```
You mustn't pass the version back. Instead, create a constexpr uint32_t kVersion = "your version" and test it against passed version. If the version is suitable, create the object.

msisov commented on ui/ozone/platform/wayland/wayland_connection.h:
```diff
@@ -110,6 +110,14 @@ class WaylandConnection : public PlatformEventSource,
       ClipboardDelegate::GetMimeTypesClosure callback) override;
   bool IsSelectionOwner() override;
 
+  // Requests the dragged data to system.
```
Can you please add a more descriptive comment to each of the methods?

msisov commented on ui/ozone/platform/wayland/wayland_connection.h:
```diff
@@ -196,6 +204,7 @@ class WaylandConnection : public PlatformEventSource,
   std::unique_ptr<WaylandDataDeviceManager> data_device_manager_;
   std::unique_ptr<WaylandDataDevice> data_device_;
   std::unique_ptr<WaylandDataSource> data_source_;
```
is it really necessary to have two WaylandDataSources?

jkim-julie replied:
I think CopyAndPaste and Dnd could work at the same time. For instance, a user copies something, clipboard holds it and he could try DnD. When 'WaylandDataSource::OnCancel' or 'WaylandDataSource::OnDnDFinished' can not destroy the instance because clipboard holds copied data yet. I also checked Weston code and it manages each data source for DnD, editor and terminal.

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -39,29 +70,109 @@ void WaylandDataDevice::RequestSelectionData(const std::string& mime_type) {
-  std::string contents;
+void WaylandDataDevice::ReadDataFromFD(std::string* contents,
+                                       base::ScopedFD fd,
+                                       const std::string& mime_type) {
```
check |contents| is a valid pointer.

jkim-julie replied:
Done

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -39,29 +70,109 @@ void WaylandDataDevice::RequestSelectionData(const std::string& mime_type) {
+                                          const std::string& mime_type) {
+  std::string contents;
+  ReadDataFromFD(&contents, std::move(fd), mime_type);
+  LOG(ERROR) << "contents:" << contents;
```
remove LOG

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -39,29 +70,109 @@ void WaylandDataDevice::RequestSelectionData(const std::string& mime_type) {
+                                     const std::string& mime_type) {
+  std::string contents;
+  ReadDataFromFD(&contents, std::move(fd), mime_type);
+  LOG(ERROR) << __FUNCTION__ << ", " << mime_type << ", contents:" << contents;
```
remove log

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -39,29 +70,109 @@ void WaylandDataDevice::RequestSelectionData(const std::string& mime_type) {
+
+  wl_data_device_start_drag(data_device_.get(), data_source, surface,
+                            icon_surface_.get(), connection_->serial());
+  wl_display_flush(connection_->display());
```
use connection_->ScheduleFlush() instead.

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -108,4 +298,62 @@ void WaylandDataDevice::SyncCallback(void* data,
+void WaylandDataDevice::CreateDragImage(const SkBitmap& bitmap) {
+  gfx::Size size(bitmap.width(), bitmap.height());
+  if (!CreateSHMBuffer(size))
+    return;
```
nit: space after return

jkim-julie replied:
Done

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -108,4 +298,62 @@ void WaylandDataDevice::SyncCallback(void* data,
   data_device->sync_callback_.reset();
 }
 
+// TODO: Make it common API as it's duplicate of WaylandCursor::CreateSHMBuffer.
```
TODO(jkim).

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -108,4 +298,62 @@ void WaylandDataDevice::SyncCallback(void* data,
+bool WaylandDataDevice::CreateSHMBuffer(const gfx::Size& size) {
+  shm_ = connection_->shm();
+
+  if (size == size_)
```
can you explain this and add a comment here?

msisov commented on ui/ozone/platform/wayland/wayland_data_device.h:
```diff
@@ -11,36 +11,82 @@
  private:
-  void ReadClipboardDataFromFD(base::ScopedFD fd, const std::string& mime_type);
+  // Registers display sync callback. Once it's called, it's reset.
+  void RegisterSyncCallback();
```
nit:space

jkim-julie replied:
Done.

msisov commented on ui/ozone/platform/wayland/wayland_data_offer.h:
```diff
@@ -42,20 +42,36 @@ class WaylandDataOffer {
 
   wl::Object<wl_data_offer> data_offer_;
   std::vector<std::string> mime_types_;
+  // actions offered by the data source
```
Actions, Action

jkim-julie replied:
It's a bitfield variable and could have multiple actions. https://people.freedesktop.org/~whot/wayland-doxygen/wayland/Client/structwl__data__offer__listener.html#ab0d4fe2ca5491f552bb8f953427eb763

msisov replied:
I meant that you should do s/actions/Actions/ and s/action/Action/

msisov commented on ui/ozone/platform/wayland/wayland_data_source.cc:
```diff
@@ -39,11 +42,31 @@ void WaylandDataSource::UpdataDataMap(
                                  wl_data_source* source,
                                  const char* mime_type) {
   NOTIMPLEMENTED();
+  if (!mime_type)
```
Do you really need this check and LOG?

msisov commented on ui/ozone/platform/wayland/wayland_data_source.cc:
```diff
@@ -67,7 +96,27 @@ void WaylandDataSource::OnSend(void* data,
+  }
+}
+
+void WaylandDataSource::OnDnDDropPerformed(void* data, wl_data_source* source) {
```
NOTIMPLEMENTED?

jkim-julie replied:
Done.

msisov commented on ui/ozone/platform/wayland/wayland_data_source.cc:
```diff
@@ -67,7 +96,27 @@ void WaylandDataSource::OnSend(void* data,
+void WaylandDataSource::OnAction(void* data,
+                                 wl_data_source* source,
+                                 uint32_t dnd_action) {
+  LOG(INFO) << __FUNCTION__ << ", the selected action :" << dnd_action;
```
nit: s/LOG..../NOTIMPLEMENTED

msisov commented on ui/ozone/platform/wayland/wayland_data_source.h:
```diff
@@ -39,6 +41,10 @@ class WaylandDataSource {
+  void Offer(std::vector<std::string>& mime_types);
+  void SetAction(int operation);
+
+  wl_data_source* data_source() { return data_source_.get(); }
```
const?

msisov commented on ui/ozone/platform/wayland/wayland_object.cc:
```diff
@@ -40,6 +40,14 @@ void delete_touch(wl_touch* touch) {
 }
 
+void delete_data_device(wl_data_device* data_device) {
+  if (wl_data_device_get_version(data_device) >=
```
conditionals with more than one line require square brackets "{}"

jkim-julie replied:
Done.

msisov commented on ui/platform_window/platform_window.h:
```diff
@@ -72,6 +73,16 @@ class PlatformWindow {
   virtual void StartWindowMoveOrResize(int hittest,
                                        gfx::Point pointer_location) = 0;
 
+  // DnD
```
this comment looks ugly :)

msisov commented on ui/platform_window/platform_window.h:
```diff
@@ -72,6 +73,16 @@ class PlatformWindow {
                                        gfx::Point pointer_location) = 0;
 
+  // DnD
+  // TODO: Make pure virtual functions.
```
TODO(jkim)

msisov replied:
the same below

msisov commented on ui/platform_window/platform_window_delegate.h:
```diff
@@ -55,6 +58,38 @@ class PlatformWindowDelegate {
 
   virtual void OnActivationChanged(bool active) = 0;
+
+  // Drag and Drop
```
doesn't look good..again :)

msisov commented on ui/views/widget/desktop_aura/desktop_drag_drop_client_ozone.cc:
```diff
@@ -0,0 +1,379 @@
+
+  platform_window_->StartDrag(mime_types, operation, drag_image);
+  run_loop.Run();
+  return false;
```
return false? oO

msisov commented on ui/views/widget/desktop_aura/desktop_drag_drop_client_ozone.cc:
```diff
@@ -0,0 +1,379 @@
+void DesktopDragDropClientOzone::OnDragDataReceived(
+    const std::string& contents) {
+  if (!data_collector_) {
+    LOG(ERROR) << "Completed the drop session already.";
```
nit: The DnD session has already been completed.

msisov commented on ui/views/widget/desktop_aura/desktop_drag_drop_client_ozone.h:
```diff
@@ -0,0 +1,168 @@
+    std::list<std::string> unprocessed_mime_types_;
+  };
+
+ public:
```
AFAIK, chromium style disallows having 
public:
private:
public:

Can you use forward declaration instead and move the class above into the private section bellow?

msisov commented on ui/views/widget/desktop_aura/desktop_window_tree_host_platform.cc:
```diff
@@ -117,9 +117,10 @@ DesktopWindowTreeHostPlatform::CreateTooltip() {
-  return nullptr;
+  drag_drop_client_ = new DesktopDragDropClientOzone(window(), cursor_manager,
+                                                     platform_window());
+  // drag_drop_client_->Init();
```
remove this

msisov commented on ui/views/widget/desktop_aura/desktop_drag_drop_client_ozone.cc:
```diff
@@ -0,0 +1,379 @@
+  }
+
+  point_.SetPoint(x, y);
+  serial_ = serial;
```
it doesn't look like you need a serial_ at all here. it's wayland specific and it mustn't be passed to the ozone subsystem.

jkim-julie commented:
I updated the patch with addressing review comment. PTAL.

msisov commented:
Some side questions - will DesktopDragDropClientOzone work with other platforms as well? Have you checked it for consistency with the X11 backend, for example?

jkim-julie commented:
Not yet. I was thinking of implementing X11 backend after finishing Wayland backend. Maybe it could cause interface or design change. Then, may I send PR after confirm that X11 backend works? If we don't need to hurry, I prefer to get a review after finishing X11 backend as well.

msisov commented:
Yes, please. Otherwise, the patch looks fine to me

jkim-julie commented:
@msisov, PTAL. It includes DnD on X11 as well. It created X11DragSource and X11DragContext. They are platform related parts from DesktopDragDropClientAuraX11 and X11WindowOzone catches drag events. In order to reuse OSExchangeDataProviderAuraX11, it introduces OSExchangeDataProviderAuraX11Base and some code only related to platform events is in OSExchangeDataProviderAuraX11.

jkim-julie committed:
Implements Drag and Drop for Ozone/Wayland

It introduces DesktopDragDropClientOzone and some APIs related to DnD
to PlatformWindowDelegate to deliver DnD status from platform and
system. OnDragEnter/OnDragMotion/OnDragDrop/OnDragLeave reports DnD
status with information based on the status. OnDragDataReceived is
called when it gets dragged data from system, OnDragDataRequest is
called when system requests dragged data and OnCloseDrag is called
when DnD is canceled or performed. These APIs are connected to
DesktopDragDropClientOzone.

It exteneds WaylandDataDevice, WaylandDataSource and WaylandDataOffer
for DnD on Wayland.

The working sequence is blow.
1) When it gets dragging.
WaylandDataDevice gets OnEnter() and DesktopDragDropClientOzone set
dragged data if it has |source_data_|. It means dragging is started
internally. After that, it continously gets Motion and Drop events.
When it gets OnDragDrop, it tries to read dragged data if it doesn't
have yet. As OnDragLeave event comes right after OnDragDrop, it
defers OnDragLeave event handling. After finishing it, it calls
callback and OnDragLeave is handled. The reason that it reads
dragged data on OnDragDrop event, it's blocked if |fd| offered by
|drag_offer_| doesn't have data yet.

2) When it sends dragging.
It calls 'wl_data_device_start_drag' through
'DesktopDragDropClientOzone::StartDragAndDrop'. After getting
OnDragEnter/Motion, it gets WaylandDataSource::OnSend() and fills
|fd| with the dragged data. If Drop is performed without problem,
it gets OnDnDDropPerformed() and OnDnDFinished(). Otherwise,
it gets OnCancel().

jkim-julie committed:
Implements Drag and Drop for Ozone/X11

It introduces X11DragSource and X11DragContext to handle DnD
on Ozone/X11. They are from DesktopDragDropClientAuraX11 which
includes implementation for platform layer.

X11DragSource handles event and deliver the data at a server side
and X11DragContext handles events and read the data at a client
side.

It splits OSExchangeDataProviderAuraX11 to OSExchangeDataProviderAuraX11
and OSExchangeDataProviderAuraX11Base to resue it.
OSExchangeDataProviderAuraX11 keeps only platform event specific part.

msisov commented on ui/ozone/platform/wayland/wayland_connection.h:
```diff
@@ -104,6 +104,8 @@ class WaylandConnection : public PlatformEventSource,
   // Returns the current pointer, which may be null.
   WaylandPointer* pointer() { return pointer_.get(); }
 
+  WaylandDataSource* drag_data_source() { return drag_data_source_.get(); }
```
WaylandDataSource* drag_data_source() const?

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -39,29 +105,128 @@ void WaylandDataDevice::RequestSelectionData(const std::string& mime_type) {
+void WaylandDataDevice::RequestDragData(
+    const std::string& mime_type,
+    base::OnceCallback<void(const std::string&)> callback) {
+  base::ScopedFD fd = drag_offer_->Receive(mime_type);
```
can it happen that drag_offer_ is not available by this time? What's about having
while(!drag_offer_)
  wl_display_roundtrip(...)?

jkim-julie replied:
I didn't know that Wayland has that API. Good to know. Thanks but on this point, RequestDragData() is triggered by OnDrop event and |drag_offer_| is created by OnEnter event. If |drag_offer_| is not available in RequestDragData(), it means that OnDrop event happens without OnEnter event. I think it's important to get correct event sequence from platform for DnD and we can't make DnD work even if we wait for it here when event sequence is wrong and |drag_offer_| is invalid. WDYT?

jkim-julie replied:
And I removed |window_| validation checking because |window_| is from 'wl_surface_get_user_data(surface)' on DragEnter event. if the surface has invalid window data, DnD will not work even if WaylandDataDevice gets properly all events such as OnEnter, OnMotion, OnDrop and OnLeave. So, I added window checking to each event listener before take chromium event logic. WDYT?

msisov replied:
ok.

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -39,29 +105,128 @@ void WaylandDataDevice::RequestSelectionData(const std::string& mime_type) {
+  }
+
+  if (!window_) {
+    LOG(ERROR) << "Could not find window.";
```
maybe, there is no sense in calling drag_offer_->Receive if there is no window_.

jkim-julie commented on ui/ozone/platform/wayland/wayland_connection.h:
```diff
@@ -104,6 +104,8 @@ class WaylandConnection : public PlatformEventSource,
   // Returns the current pointer, which may be null.
   WaylandPointer* pointer() { return pointer_.get(); }
 
+  WaylandDataSource* drag_data_source() const { return drag_data_source_.get(); }
```
I added 'const' here as @msisov commented.

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -5,29 +5,95 @@
 
+namespace {
+
+auto& kMimeTypeText = Clipboard::kMimeTypeText;
```
this look weird.

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -5,29 +5,95 @@
+auto& kMimeTypeText = Clipboard::kMimeTypeText;
+constexpr char kMimeTypeTextUTF8[] = "text/plain;charset=utf-8";
+
+const uint32_t kShmFormat = WL_SHM_FORMAT_ARGB8888;
```
constexpr

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -5,29 +5,95 @@
+constexpr char kMimeTypeTextUTF8[] = "text/plain;charset=utf-8";
+
+const uint32_t kShmFormat = WL_SHM_FORMAT_ARGB8888;
+const SkColorType kColorType = kBGRA_8888_SkColorType;
```
constexpr

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -5,29 +5,95 @@
+  return operation;
+}
+
+void AddStringToOSExchangeData(OSExchangeData* os_exchange_data,
```
chromium style requires const values to be passed first and then pointer to data that it going to be changed.

msisov commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -5,29 +5,95 @@
+  os_exchange_data->SetString(string16);
+}
+
+void AddToOSExchangeData(OSExchangeData* os_exchange_data,
```
same here.

jkim-julie commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -5,29 +5,94 @@
+}
+
+void AddStringToOSExchangeData(const std::string& data,
+                               OSExchangeData* os_exchange_data) {
```
Changed the parameter order.

jkim-julie commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -5,29 +5,94 @@
+constexpr char kMimeTypeTextUTF8[] = "text/plain;charset=utf-8";
+
+constexpr uint32_t kShmFormat = WL_SHM_FORMAT_ARGB8888;
+constexpr SkColorType kColorType = kBGRA_8888_SkColorType;
```
Used constexpr instead of const.

jkim-julie commented on ui/ozone/platform/wayland/wayland_data_device.cc:
```diff
@@ -5,29 +5,94 @@
 
+namespace {
+
+constexpr char kMimeTypeText[] = "text/plain";
```
Updated mimetype definition.
